File: Tracer/Sphere.cpp
#include "Sphere.h"
#include "Ray.h"

using namespace Tracer;

Vector Sphere::getNormal(const Vector& p) const
{
	return (p - centre).unit();
}

Sphere::Sphere(const Vector& center, const float radius, const Material& material) 
	: centre(center), radius(radius), ISurface(material) { }

bool Sphere::intersect(const Ray& r, Intersection& out)
{
	float a, b, c, det;
	Vector l = r.o - centre;

	a = r.d * r.d;
	b =  2 * (r.d * l);
	c = l * l - radius * radius;
	det = b * b - 4 * a * c;

	if (det < 0)
		return false;

	float t = (-b - det) / (2 * a);
	out.point = r(t);
	out.normal = getNormal(out.point);
	out.m = getMaterial();

	return true;
}

File: src/tde/sphere.rs
use crate::rendering::Intersectable;
use crate::math::Vector;
use super::Ray;

pub struct Sphere {
    pub c: Vector,
    pub r: f64
}

impl Intersectable for Sphere {
    fn intersect(&self, r: &Ray) -> Option<f64> {
        let  (a, b, c, det): (f64,f64,f64,f64);
        let l: Vector = r.o - self.c;
        a = r.d * r.d;
        b = 2.0 * (r.d * l);
        c = l * l - self.r * self.r;
        det = b * b - 4.0 * a * c;

        if det < 0.0 {
            return None;
        }

        let t: f64 = (-b - det.sqrt()) / (2.0 * a);
        if t < 0.01 {
            return None;
        }

        return Some(t);
    }

    fn get_normal(&self, p: Vector) -> Vector {
        return (p - self.c).unit();
    }
}
